### PR TITLE
Adding retry logic to cache fallback, handling fallback on cache calls

### PR DESCRIPTION
**Pull request by EricTendian**
## Summary of Changes

This PR does a few things:
- updates the package to work with Laravel 5.5+
- adds retry logic to the resolve() function
- introduces error handling/fallbacks and retries to any cache calls
- improves config file documentation
- gets all tests to pass w/ newest PHPUnit

## Testing

- [ ] Functional tests
- [x] Unit tests
- [x] Manual tests
- [ ] Other (describe...)

New tests were written to handle the __call cases, and other tests were updated. All flows are now covered. I've also verified this manually in one of our apps.

## PR Checklist

- [x] My code adheres to [Packback's code styling and standards](https://sites.google.com/a/packback.co/engineering-process/engineering-technical-docs/code-style-guides). If not, add an explanation for why it doesn't.
- [x] I have selected at least one primary assignee for this PR. All of these people need to approve the pull request before it is merged.
- [x] I have selected any number of reviewers for this pull request.  Reviewers are people who should know about the pull request, but whose sign-off is not a blocker to merging.
- [ ] If this is a BUG that was found outside of development, I have added some text describing how it got through our tests, and how we will prevent similar issues in the future.

### CODE REVIEW

File: src/config/cache_fallback.php [hunk 1]
```diff
@@ -1,12 +1,66 @@
...
+
     'fallback_order' => [
         'redis',
         'memcached',
         'database',
         'cookie',
         'file',
         'array'
-    ]
+    ],
```
Comment (dbhynds) [inline src/config/cache_fallback.php:23]: In terms of the order is this something we can somehow configure via envs? How would you envision doing that?

Comment (EricTendian) [reply src/config/cache_fallback.php:23]: Yeah, we'll be publishing the config in our own app, and then from there we can override settings - this is just the default if no envs are used.

**Comment by dbhynds** [comment]
The code looks fine, and the tests look reasonable. You said you manually tested this. Can you talk me through how you manually tested it briefly? Was it just like... have the app running, stop redis, see that it doesn't die, restart redis?

**Comment by dbhynds** [comment]
Oh also... a lot of our caching depends on cache tags, but not all cache drivers support cache tags. What happens if we fallback on a driver that doesn't support the same functionality as the previous driver?

**Comment by EricTendian** [comment]
@dbhynds I'll show you in a subsequent PR how this was tested, but essentially it was having a loop running which was constantly pulling a value from the cache, then shutting down Redis, seeing that it didn't throw an exception, and then starting Redis back up, and seeing it go back to the previous value. Regarding cache tags, I tried it with the file driver and no exception occurred. I don't think we should worry too much about that since the main idea with this package is to have a graceful failure so we do database queries to get the relevant data, instead of simply giving a 500 error.

**Review by dbhynds — APPROVED** [review]

**Pull request merged**